pilot/pkg/networking/core/v1alpha3/route/route.go

```diff
@@ -101,6 +107,39 @@ func BuildSidecarVirtualHostWrapper(routeCache *Cache, node *model.Proxy, push *
 ) []VirtualHostWrapper {
 	out := make([]VirtualHostWrapper, 0)
 
+	// For correctness, we need to maintain a mapping of hostnames to the
+	// virtual services that have the most specific wildcard host match
+	// match for that hostname. Unfortunately, we have to loop through all of
+	// hosts in all of the virtualservices an extra time...
+	//
+	// TODO: Investigate whether it's worth it to maintain this map in the
+	// pushContext or something else higher level for _all_ virtualServices
+	vsWithMostSpecificWildcardHosts := make(map[host.Name]vsHostEntry)
+	for _, vs := range virtualServices {
```

Comment:
What is impact of performance on looping through all virtual services every time?

Reply:
I'm glad someone asked; I left a comment calling this out. It's not as bad as it could be since we only need to process hosts with wildcards; I don't anticipate their being even 100 such hosts/virtualServices for a given pilot instance. That being said, I'm definitely open to optimizations as long as we can avoid unneeded complexity. The egress listener could maintain this index at its level so we don't have to recompute so frequently. The extreme version would be to compute a true for all of the wildcards (unnecessary IMO).

Reply:
@hzxuzhonghu the loop can't be moved down because we need the full map computed in the next loop

Reply:
Yeah, it should be precalculated beforehand. The performance influence is unneglected.  Here it does calculate on time per route. 
It can be optimized maybe one time per proxy or even one time per Sidecar.

Reply:
Ack, I figured as much but wanted a second opinion. I'll add the map to the egress listener